logging/context.go

```diff
@@ -0,0 +1,78 @@
+package logging
+
+import (
+	"context"
+
+	"go.viam.com/utils"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type debugLogKeyType int
+
+const debugLogKeyID = debugLogKeyType(iota)
+
+// EnableDebugMode returns a new context with debug logging state attached. An empty `debugLogKey`
+// generates a random value.
+func EnableDebugMode(ctx context.Context, debugLogKey string) context.Context {
+	if debugLogKey == "" {
+		debugLogKey = utils.RandomAlphaString(6)
+	}
+	return context.WithValue(ctx, debugLogKeyID, debugLogKey)
+}
+
+// IsDebugMode returns whether the input context has debug logging enabled.
+func IsDebugMode(ctx context.Context) bool {
+	return GetName(ctx) != ""
+}
+
+// GetName returns the debug log key included when enabling the context for debug logging.
+func GetName(ctx context.Context) string {
+	valI := ctx.Value(debugLogKeyID)
+	if val, ok := valI.(string); ok {
+		return val
+	}
+
+	return ""
+}
+
+const dtNameMetadataKey = "dtName"
+
+// UnaryClientInterceptor adds debug directives from the current context (if any) to the
+// outgoing request's metadata.
+func UnaryClientInterceptor(
+	ctx context.Context,
+	method string,
+	req, reply interface{},
+	cc *grpc.ClientConn,
+	invoker grpc.UnaryInvoker,
+	opts ...grpc.CallOption,
+) error {
+	if IsDebugMode(ctx) {
+		ctx = metadata.AppendToOutgoingContext(ctx, dtNameMetadataKey, GetName(ctx))
+	}
+
+	return invoker(ctx, method, req, reply, cc, opts...)
+}
+
+// UnaryServerInterceptor creates a new operation in the current context before passing
+// it to the unary response handler. If the incoming RPC metadata contains a directive for
+// distributed tracing, attach that to the context.
```

Comment:
where are we creating a new operation? did you mean to say we are creating a new "context"?

Reply:
Err yeah. Nice catch. At MDB the word for a context was an `OperationContext`. And I just flubbed the two here.

Reply:
That said, the comment is weird. I redid it.

Reply:
> At MDB the word for a context was an OperationContext.

Gotcha - intuitively it made sense, but I was just concerned about conflating this with [our robot operation concept](https://github.com/viamrobotics/rdk/blob/6155573880b7c3b56988f46385526d4294464a8f/operation/opid.go#L27)